launch/launch/launch_context.py

```diff
@@ -169,10 +169,26 @@ def _push_environment(self):
     def _pop_environment(self):
         if not self.__environment_stack:
             raise RuntimeError('environment stack unexpectedly empty')
-        os.environ = self.__environment_stack.pop()
+
+        # Note that we cannot just assign os.environ to the copy of the
+        # environment that we saved during _push_environment() above.
+        # That's because os.environ.copy() returns a dict, while os.environ
+        # is a os._Environ object.  They act the same in many, but not all
+        # circumstances.  Instead, when we go to pop the environment, we
+        # clear out the current environment and push the values from the
+        # copy one-by-one into os.environ.
+
+        old_env = self.__environment_stack.pop()
+        os.environ.clear()
+        for k, v in old_env.items():
+            os.environ[k] = v
```

Comment:
nit, I think after `clear()` this for loop could be replaced with `os.environ.update(old_env)`.

Reply:
[1d690bd](https://github.com/ros2/launch/pull/656/commits/1d690bd7848baabad3f9f823f3847e7f3c687b31)